components/validator/guardian.go

```diff
@@ -46,59 +67,289 @@ func NewGuardian(cfg Config, l log.Logger) (*Guardian, error) {
+				g.log.Info("guardian: need to wait for L2 blocks sync completed", "currentL2", currentL2, "latestOutputL2", latestOutputL2)
+				continue
+			}
+
+			// currentL1 and creationEndedL1 are used for searching events of ReadyToProve in L1 blocks
+			currentL1 := new(big.Int).SetUint64(status.CurrentL1.Number)
+			creationPeriodL1 := new(big.Int).Div(g.creationPeriodSeconds, big.NewInt(12))
+			creationPeriodL2 := new(big.Int).Div(g.creationPeriodSeconds, g.l2BlockTime)
+			if currentL1.Cmp(creationPeriodL1) != 1 || currentL2.Cmp(creationPeriodL2) != 1 {
+				g.log.Warn("guardian: there is no output when the creation period is over yet", "currentL1", currentL1, "currentL2", currentL2, "creationPeriodL1", creationPeriodL1, "creationPeriodL2", creationPeriodL2)
+				continue
+			}
+
+			creationEndedL1 := new(big.Int).Sub(currentL1, creationPeriodL1)
+			g.updateInspectionWindow(creationEndedL1, currentL1)
+
+			creationEndedL2 := new(big.Int).Sub(currentL2, creationPeriodL2)
+			// finalizedL2 and creationEndedL2 is used to get outputIndex whose creation period is ended but not finalized
+			finalizedL2 := new(big.Int).Sub(currentL2, new(big.Int).Div(g.finalizationPeriodSeconds, g.l2BlockTime))
+			finalizedL2 = math.BigMax(common.Big1, finalizedL2)
+
+			// if g.checkpoint is nil, scan all the outputs whose creation period is ended but not finalized
+			if g.checkpoint == nil {
+				func() {
+					cCtx, cCancel := context.WithTimeout(g.ctx, g.cfg.NetworkTimeout)
+					defer cCancel()
+					startOutputIndex, err := g.l2ooContract.GetL2OutputIndexAfter(utils.NewSimpleCallOpts(cCtx), finalizedL2)
+					if err != nil {
+						g.log.Error("guardian: failed to get output index after", "err", err, "afterL2Block", finalizedL2.Uint64())
+						return
+					}
+
+					cCtx, cCancel = context.WithTimeout(g.ctx, g.cfg.NetworkTimeout)
+					defer cCancel()
+					endOutputIndex, err := g.l2ooContract.GetL2OutputIndexAfter(utils.NewSimpleCallOpts(cCtx), creationEndedL2)
+					if err != nil {
+						g.log.Error("guardian: failed to get output index after", "err", err, "afterL2Block", creationEndedL2.Uint64())
+						return
+					}
+
+					for i := startOutputIndex; i.Cmp(endOutputIndex) != 1; i.Add(i, common.Big1) {
+						g.wg.Add(1)
+						go g.inspectOutput(new(big.Int).Set(i))
+					}
+
+					g.checkpoint = endOutputIndex
+				}()
+			} else {
+				func() {
+					cCtx, cCancel := context.WithTimeout(g.ctx, g.cfg.NetworkTimeout)
+					defer cCancel()
+					outputIndex, err := g.l2ooContract.GetL2OutputIndexAfter(utils.NewSimpleCallOpts(cCtx), creationEndedL2)
+					if err != nil {
+						g.log.Error("guardian: failed to get output index after", "err", err, "afterL2Block", finalizedL2.Uint64())
+						return
+					}
+
+					for i := new(big.Int).Add(g.checkpoint, common.Big1); i.Cmp(outputIndex) != 1; i.Add(i, common.Big1) {
+						g.wg.Add(1)
+						go g.inspectOutput(outputIndex)
+					}
+
+					g.checkpoint = outputIndex
```

Comment:
Shouldn't we only check the output up to `outputIndex - 1` and update checkpoint with `outputIndex - 1`? We get `outputIndex` using `GetL2OutputIndexAfter`, so it means `outputIndex` can be in creation period yet.